benchmarks/halo_finder/ArborX_HaloFinder.hpp

```diff
@@ -0,0 +1,264 @@
+/****************************************************************************
+ * Copyright (c) 2012-2020 by the ArborX authors                            *
+ * All rights reserved.                                                     *
+ *                                                                          *
+ * This file is part of the ArborX library. ArborX is                       *
+ * distributed under a BSD 3-clause license. For the licensing terms see    *
+ * the LICENSE file in the top-level directory.                             *
+ *                                                                          *
+ * SPDX-License-Identifier: BSD-3-Clause                                    *
+ ****************************************************************************/
+
+#ifndef ARBORX_HALO_FINDER_HPP
+#define ARBORX_HALO_FINDER_HPP
+
+#include <ArborX_DetailsSortUtils.hpp>
+#include <ArborX_DetailsUtils.hpp>
+#include <ArborX_LinearBVH.hpp>
+#include <ArborX_Macros.hpp>
+
+#include <chrono>
+
+namespace ArborX
+{
+
+template <typename View>
+struct Wrapped
+{
+  View _M_view;
+  double _r;
+};
+
+template <typename View>
+auto wrap(View v, double r)
+{
+  return Wrapped<View>{v, r};
+}
+
+namespace Traits
+{
+template <typename View>
+struct Access<Wrapped<View>, PredicatesTag>
+{
+  using memory_space = typename View::memory_space;
+  static size_t size(Wrapped<View> const &w) { return w._M_view.extent(0); }
+  static KOKKOS_FUNCTION auto get(Wrapped<View> const &w, size_t i)
+  {
+    return attach(intersects(Sphere{w._M_view(i), w._r}), (int)i);
+  }
+};
+} // namespace Traits
+
+namespace HaloFinder
+{
+template <typename MemorySpace>
+struct CCSCallback
+{
+  using tag = ArborX::Details::InlineCallbackTag;
+  Kokkos::View<int *, MemorySpace> stat_;
+
+  KOKKOS_INLINE_FUNCTION
+  int representative(int const i) const
+  {
+    // ##### ECL license (see LICENSE.ECL) #####
+    int curr = stat_(i);
+    if (curr != i)
+    {
+      int next, prev = i;
+      while (curr > (next = stat_(curr)))
+      {
+        stat_(prev) = next;
+        prev = curr;
+        curr = next;
+      }
+    }
+    return curr;
+  }
+
+  template <typename Query, typename Insert>
+  KOKKOS_FUNCTION void operator()(Query const &query, int j,
+                                  Insert const &) const
+  {
+    int const i = ArborX::getData(query);
+
+    // only process edge in one direction
+    if (i > j)
+    {
+      // initialize to the first neighbor that's smaller
+      if (Kokkos::atomic_compare_exchange(&stat_(i), i, j) == i)
+        return;
```

Comment:
This is the init stage. Need to think more carefully whether this `return` is warranted, or if it makes sense to continue here. It's all about races between threads.